### ISSUE TITLE
Shape change during TPU training

### ISSUE THREAD

**Issue by maxima120** [tensorflow/tensorflow#29825]
Once I start train my model on TPU v3-8 I get this output and model start recompiling and continue train with different shape:
```
INFO:tensorflow:New input shapes; (re-)compiling: mode=train (# of cores 8), [TensorSpec(shape=(128,), dtype=tf.int32, name='core_id0'), TensorSpec(shape=(128, 1024, 7), dtype=tf.float32, name='lstm_input_10'), TensorSpec(shape=(128, 4), dtype=tf.float32, name='dense_target_30')]
INFO:tensorflow:Overriding default placeholder.
INFO:tensorflow:Remapping placeholder for lstm_input
INFO:tensorflow:Started compiling
INFO:tensorflow:Finished compiling. Time elapsed: 5.74567985534668 secs
INFO:tensorflow:Setting weights on TPU model.
 9216/11126 [=======================>......] - ETA: 4s - loss: nan - categorical_accuracy: 0.0250INFO:tensorflow:New input shapes; (re-)compiling: mode=train (# of cores 8), [TensorSpec(shape=(110,), dtype=tf.int32, name='core_id0'), TensorSpec(shape=(110, 1024, 7), dtype=tf.float32, name='lstm_input_10'), TensorSpec(shape=(110, 4), dtype=tf.float32, name='dense_target_30')]
INFO:tensorflow:Overriding default placeholder.
INFO:tensorflow:Remapping placeholder for lstm_input
INFO:tensorflow:Started compiling
INFO:tensorflow:Finished compiling. Time elapsed: 8.460586786270142 secs
```

Compare shapes before train kicks in and after:
```
INFO:tensorflow:New input shapes; (re-)compiling: mode=train (# of cores 8), [TensorSpec(shape=(128,), dtype=tf.int32, name='core_id0'), TensorSpec(shape=(128, 1024, 7), dtype=tf.float32, name='lstm_input_10'), TensorSpec(shape=(128, 4), dtype=tf.float32, name='dense_target_30')]
INFO:tensorflow:New input shapes; (re-)compiling: mode=train (# of cores 8), [TensorSpec(shape=(110,), dtype=tf.int32, name='core_id0'), TensorSpec(shape=(110, 1024, 7), dtype=tf.float32, name='lstm_input_10'), TensorSpec(shape=(110, 4), dtype=tf.float32, name='dense_target_30')]
```
Model shapes:
```
Layer (type)                 Output Shape              Param #   
=================================================================
lstm_input (InputLayer)      (None, 1024, 7)           0         
_________________________________________________________________
lstm (LSTM)                  (None, 1024, 256)         270336    
_________________________________________________________________
dropout (Dropout)            (None, 1024, 256)         0         
_________________________________________________________________
lstm_1 (LSTM)                (None, 256)               525312    
_________________________________________________________________
dropout_1 (Dropout)          (None, 256)               0         
_________________________________________________________________
dense (Dense)                (None, 4)                 1028      
=================================================================
```

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Debian 9
- TensorFlow installed from (source or binary): 1.13.1
- Python version: 3.5
- GPU model and memory: TPU v3-8

**Describe the current behavior**

shapes change from 128 to 110 during first few batches training

**Describe the expected behavior**

shapes don't change

**Comment by gadagashwini** [comment]
@maxima120 In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

**Comment by maxima120** [comment]
I will need to send you data and code for this. I don't think you can use just a snippet. Please let me know how to contact you. Thanks

**Comment by gadagashwini** [comment]
@maxima120 Will it be possible you to share code and data in zip file. If it is not private. Thanks!

**Comment by maxima120** [comment]
code is no secret but data I don't want to put on the open. hope you understand.

**Comment by maxima120** [comment]
```
import tensorflow as tf
import keras

window_size = 1024
n_steps = len(data[0]) - window_size
inputs_n = 7
outputs_n = 4
neurons = 128
learning_rate = 0.00001
activation = 'softmax'

from keras.layers import Dense, Activation, Dropout, LSTM, Dropout
from tensorflow.keras import layers

tf.keras.backend.clear_session()
tf.keras.backend.get_session().run(tf.global_variables_initializer())

model = tf.keras.Sequential()
model.add(layers.LSTM(neurons, input_shape=(window_size, inputs_n), return_sequences=True)) 
model.add(layers.LSTM(neurons))
model.add(layers.Dense(neurons, activation='relu'))
model.add(layers.Dropout(0.2))
model.add(layers.Dense(outputs_n, activation=activation))

opt = tf.train.RMSPropOptimizer(learning_rate)

model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['categorical_accuracy'])

print(model.summary())

import time
# loop by day - separate features and labels

buy_sell = 'sell'
print('Training.. LABEL:', buy_sell, 'neurons:', neurons, 'learning rate:', learning_rate, 'activation:', activation)

epochs_n = 1
epochs =  range(epochs_n)

histories = list()
day_count = 0

start_time = time.time()

# d is tuple from groupby - d[0] = date, d[1] = values
for epoch in epochs:
    for d in data : 
        # get arrays for the day
        features = np.asarray(d)[:,2:9].astype(dtype = 'float32')
        labels = np.asarray(d)[:, 9:13].astype(dtype = 'int8')
        
        X,y = split_sequence(features, labels, window_size)

        try:
            H = model.fit(X,y, batch_size = window_size)
            histories.append(H.history)
        except Exception as e:
            print('** train exception :', e)
            continue
        
    #for days
#for epoch

print('DONE..')

**Comment by gadagashwini** [comment]
> code is no secret but data I don't want to put on the open. hope you understand.

Thanks for sharing the code.

**Comment by maxima120** [comment]
this has everything you need to reproduce:
https://gist.github.com/maxima120/63e1cdca833264afb4192120b72d2225

**Comment by rjpower** [comment]
The re-compilation is expected.

Keras `model.fit` passes over the entire dataset once for each epoch. If the number of examples in the dataset is not divisible by the batch size, the last batch will be of a different size than the other training batches. (This works similarly on a GPU, you just don't see this message).

**Comment by maxima120** [comment]
Ok please clarify further. If say a regular batch was compiled for 128, then the last was recompiled for 110 for first epoch.. **Can I be sure that all the following epoch will use 128** for all the batched but last?

**Comment by rjpower** [comment]
Yes, this should be the standard Keras behavior: it uses the requested batch size until the last batch of each epoch.

As compilation is cached for each new batch size, you shouldn't see a recompile after the first epoch.
